### PR TITLE
gh-104683: clinic.py: Modernise `parse_converter()` using pattern-matching

### PR DESCRIPTION
**Pull request by AlexWaygood**
And add type hints while we're here

* Issue: gh-104683

### CODE REVIEW

**Review by sunmy2019 — APPROVED** [review]

**Review by erlend-aasland — APPROVED** [review]
Neat!

File: Tools/clinic/clinic.py [hunk 1]
```diff
@@ -4906,22 +4906,27 @@ def bad_node(self, node):
...
-
-        name = annotation.func.id
-        symbols = globals()
-
-        kwargs = {node.arg: eval_ast_expr(node.value, symbols) for node in annotation.keywords}
-        return name, False, kwargs
+    KwargDict = dict[str | None, Any]
+
+    def parse_converter(
+        self, annotation: ast.expr | None
```
Comment (AlexWaygood) [inline Tools/clinic/clinic.py]: The `self` argument isn't actually used in this function, so we could possibly move this function out of the class body and into the global scope. Or we could make it a staticmethod. Or we could leave it as it is. Thoughts @erlend-aasland?

Comment (erlend-aasland) [reply Tools/clinic/clinic.py]: I'd say either static method or status quo.

**Comment by erlend-aasland** [comment]
Oh, before you land, could you please check that we've got coverage for those branches?

File: Tools/clinic/clinic.py (hunk 1 above)
Comment (erlend-aasland) [reply Tools/clinic/clinic.py]: Let's go for static method. It'll probably be a while until we touch this again, so let's exploit this PR to make that improvement.

File: Tools/clinic/clinic.py [hunk 2]
```diff
@@ -4906,22 +4906,26 @@ def bad_node(self, node):
...
-        name = annotation.func.id
-        symbols = globals()
-
-        kwargs = {node.arg: eval_ast_expr(node.value, symbols) for node in annotation.keywords}
-        return name, False, kwargs
+    @staticmethod
+    def parse_converter(annotation: ast.expr | None) -> tuple[str, bool, KwargDict]:
+        match annotation:
+            case ast.Constant(value=str() as value):
+                return value, True, {}
```
Comment (AlexWaygood) [inline Tools/clinic/clinic.py:5025]: This branch is covered here:

https://github.com/python/cpython/blob/60993ba8b4a04dfcf2d3067d919bcce160f01bd5/Lib/test/test_clinic.py#L772-L775

_But_ there doesn't seem to be coverage for the case where an ast.Constant node is encountered that has a non-`str` value.

File: Tools/clinic/clinic.py [hunk 3]
```diff
@@ -4906,22 +4906,26 @@ def bad_node(self, node):
...
+            case ast.Name(name):
+                return name, False, {}
```
Comment (AlexWaygood) [inline Tools/clinic/clinic.py:5027]: This branch is extensively covered, by many tests, e.g.

https://github.com/python/cpython/blob/60993ba8b4a04dfcf2d3067d919bcce160f01bd5/Lib/test/test_clinic.py#L734-L741

File: Tools/clinic/clinic.py [hunk 4]
```diff
@@ -4906,22 +4906,26 @@ def bad_node(self, node):
...
+            case ast.Call(func=ast.Name(name)):
+                symbols = globals()
+                kwargs = {
+                    node.arg: eval_ast_expr(node.value, symbols)
+                    for node in annotation.keywords
+                }
+                return name, False, kwargs
```
Comment (AlexWaygood) [inline Tools/clinic/clinic.py:5034]: This branch seems well covered, by three tests:

https://github.com/python/cpython/blob/60993ba8b4a04dfcf2d3067d919bcce160f01bd5/Lib/test/test_clinic.py#L845

https://github.com/python/cpython/blob/60993ba8b4a04dfcf2d3067d919bcce160f01bd5/Lib/test/test_clinic.py#L321

https://github.com/python/cpython/blob/60993ba8b4a04dfcf2d3067d919bcce160f01bd5/Lib/test/test_clinic.py#L777

File: Tools/clinic/clinic.py [hunk 5]
```diff
@@ -4906,22 +4906,26 @@ def bad_node(self, node):
...
+            case _:
+                fail(
+                    "Annotations must be either a name, a function call, or a string."
+                )
```
Comment (AlexWaygood) [inline Tools/clinic/clinic.py:5038]: This branch seems entirely uncovered

File: Tools/clinic/clinic.py (hunk 2 above)
Comment (erlend-aasland) [reply Tools/clinic/clinic.py:5025]: Ok, can you add a test for that case?

File: Tools/clinic/clinic.py (hunk 5 above)
Comment (erlend-aasland) [reply Tools/clinic/clinic.py:5038]: Can you add a test for this branch?

Comment (AlexWaygood) [reply Tools/clinic/clinic.py:5038]: Of course! Planning on doing so in a separate PR, hence why I added "do-not-merge" :)

File: Tools/clinic/clinic.py (hunk 2 above)
Comment (AlexWaygood) [reply Tools/clinic/clinic.py:5025]: Same answer as https://github.com/python/cpython/pull/104696#discussion_r1199731051 :)

File: Tools/clinic/clinic.py (hunk 5 above)
Comment (erlend-aasland) [reply Tools/clinic/clinic.py:5038]: Great 😃

**Comment by AlexWaygood** [comment]
Test cases have been added in #104729 for the branches missing coverage.

**Pull request merged**